Enable use of data from EchoData object in `consolidate.add_depth`

emiliom opened the pull request:
Overhaul `consolidate.add_depth` to enable using data from EchoData object. Addresses #790. See the echopype documentation for [reference on several of the variables](https://echopype.readthedocs.io/en/stable/data-proc-additional.html#vertical-coordinate-z-axis-variables).

Rotations are performed using [scipy.spatial.transform.Rotation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html).

Notes and caveats:
- See the TODOs in the code.
- If both an EchoData object and external data (as scalar arguments) are passed, only data from the EchoData object is used.
- I'm seeing too many nans in the resulting `depth` variable for the sample EK80 file, so I suspect something isn't working properly.
- AZFP `tilt_x/y` variables are not handled. Only standard SONAR-netCDF4 variables are used. It assumes that tilt_x/y variables have already been transformed into pitch & roll. I'll create a new issue for this.
- I learned recently (from the EK80 reference manual) that EK80 `TransducerAlphaX/Y/Z` variables are in degrees. We're currently interpreting them as a unit vector when mapping them to `beam_direction_x/y/z`. A conversion will need to be implemented in `set_groups_ek80` (and ek60?). I'll create a new issue for this. For the sample file (below), it's not an issue b/c all `TransducerAlphaX/Y/Z` values are zero. I couldn't find an EK80 file with `TransducerAlphaX/Y/Z` values that are not all zero.
- We don't have a sample EK60/80 raw file that has all relevant variables populated with what appear to be valid values. The one file I'm using, from 2021 Hake survey ("ek80/ncei-wcsd/SH2106/EK80/Reduced_Hake-D20210701-T131621.raw"), only has some of the variables with apparently valid, non zero data: `pitch`, `roll`, and `vertical_offset`. `vertical_offset` (based on `heave`) is also suspicious b/c values are close to zero. `water_level` (0) + `vertical_offset` is supposed to be the positive distance from the ship origin.

emiliom commented:
Notes to self for creating mock `EchoData` data for testing the new functionality. Here's the minimal code needed to create a bare `EchoData` object that has just the elements needed by `consolidate.add_depth`: a `sonar_model` attribute and the "Platform" and "Sonar/Beam_group1" groups.

```python
from datatree import DataTree
import xarray as xr
from echopype.echodata.echodata import EchoData

ep = EchoData(sonar_model="EK80")

ds1 = xr.Dataset(dict(bar=("x", [1, 2])))
ds2 = xr.Dataset(dict(bar=("x", [10, 20])))
tree = DataTree.from_dict({'Platform': ds1, 'Sonar/Beam_group1': ds2}, name='root')
ep._set_tree(tree)
```

lsetiawan commented on echopype/consolidate/api.py:
```diff
@@ -51,6 +52,8 @@ def swap_dims_channel_frequency(ds: xr.Dataset) -> xr.Dataset:
 
 def add_depth(
     ds: xr.Dataset,
+    echodata: EchoData = None,
```
```suggestion
    echodata: Optional[EchoData] = None,
```

emiliom replied:
Thanks

lsetiawan commented on echopype/consolidate/api.py:
```diff
@@ -67,66 +70,183 @@ def add_depth(
-    #             )
-    # else:
-    #     tilt = 0
 
```
```suggestion
    del echodata_strict

```
Looks like pre-commit is currently not throwing error for unused argument! Interesting... anyhow, maybe it's better to just remove it here for now as a better practice: https://pycodequ.al/docs/pylint-messages/w0613-unused-argument.html#how-to-fix

emiliom replied:
Right now that argument is more for discussion; ie, whether we need something like it. If we end up seeing value in it, it will be implemented (used). Otherwise, it'll be dropped. So, I won't change anything yet.

lsetiawan commented on echopype/consolidate/api.py:
```diff
@@ -51,6 +52,8 @@ def swap_dims_channel_frequency(ds: xr.Dataset) -> xr.Dataset:
 def add_depth(
     ds: xr.Dataset,
+    echodata: EchoData = None,
+    echodata_strict=False,
```
```suggestion
    echodata_strict: bool = False,
```

emiliom commented on echopype/consolidate/api.py:
```diff
@@ -67,66 +70,183 @@ def add_depth(
         Source Sv dataset to which a depth variable will be added.
         Must contain `echo_range`.
-    depth_offset : float
+    echodata : EchoData, default=None
```
Got it. Will merge.

emiliom commented:
Added a test function to generate mock EchoData object with only the variables (and one attribute) required by `consolidate.add_depth`. Added a test that uses this mock data, together with a mock Sv Dataset.

leewujung commented:
@ctuguinay : This is also related to #1152. Let's talk through this.

leewujung commented:
We'll close this now because it'll be superseded by #1319 .
